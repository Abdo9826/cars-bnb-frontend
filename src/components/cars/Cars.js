import React, { useEffect } from 'react';
import { useNavigate } from 'react-router-dom';

const Cars = () => {
  const navigate = useNavigate();
  // #we can use navigate to redirect to another page

  useEffect(() => {
    // what is useEffect?
    // useEffect is a hook that allows us to run some code when a component is mounted or unmounted
    const token = localStorage.getItem('loginToken');
    // #we can use localStorage to store data in the browser
    // what is getItem?
    // getItem is a method that allows us to get the value of a key in localStorage

    const url = 'http://localhost:3001/cars';
    const options = {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
        // #we can use headers to set the content type
        // what is content type?
        // content type is a header that tells the server what kind of data we are sending
        Authorization: token,
        // #we can use Authorization to set the token
        // what is Authorization?

      },
    };
    fetch(url, options)
      .then((response) => response.json())
      .then((data) => console.log(data));
  }, []);

  const logout = async () => {
    const url = 'http://localhost:3001/users/sign_out';
    const options = {
      method: 'DELETE',
      headers: {
        'Content-Type': 'application/json',
        Authorization: localStorage.getItem('loginToken'),
      },
    };

    const response = await fetch(url, options);
    if (response.ok) {
      localStorage.removeItem('loginToken');
      navigate('/login');
    } else {
      throw new Error(response);
    }
    const data = await response.json();
    return data;
  };

  return (
    <div>
      <h1>Cars</h1>
      <button type="button" onClick={logout}>Logout</button>
    </div>
  );
};

export default Cars;
