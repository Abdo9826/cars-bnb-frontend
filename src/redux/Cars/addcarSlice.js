import { createAsyncThunk } from '@reduxjs/toolkit';

const addcars = createAsyncThunk(
  'cars/addcars',
  async (car, { rejectWithValue }) => {
    try {
      const token = localStorage.getItem('loginToken');
      const response = await fetch('http://localhost:3001/cars', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Authorization: token,
        },
        body: JSON.stringify({ car }),
      });
      const data = await response.json();
      return data;
    } catch (err) {
      return rejectWithValue(err);
    }
  },
);

export const deleteCars = createAsyncThunk(
  'cars/deleteCars',
  async (id, { rejectWithValue }) => {
    try {
      const token = localStorage.getItem('loginToken');
      const response = await fetch(`http://localhost:3001/cars/${id}`, {
        method: 'DELETE',
        headers: {
          'Content-Type': 'application/json',
          Authorization: token,
        },
      });
      const data = await response.json();
      console.log(response);
      return data;
    } catch (err) {
      return rejectWithValue(err);
    }
  },
);

export default addcars;
