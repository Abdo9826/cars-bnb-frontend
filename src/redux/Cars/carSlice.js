import { createSlice, createAsyncThunk, current } from '@reduxjs/toolkit';
// eslint-disable no-unused-vars

export const getCars = createAsyncThunk(
  'cars/getCars',
  async () => {
    const token = localStorage.getItem('loginToken');
    const url = 'http://localhost:3001/cars';
    const options = {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
        Authorization: token,
      },
    };
    const response = await fetch(url, options);
    const data = await response.json();
    console.log(data);
    return data;
  },
);

const carSlice = createSlice({
  name: 'cars',
  initialState: {
    cars: [],
    status: 'idle',
    error: null,
  },
  reducers: {},
  extraReducers: {
    [getCars.pending]: (state) => {
      const newState = { ...current(state) };
      newState.status = 'loading';
      return newState;
    },
    [getCars.fulfilled]: (state, action) => {
      const newState = { ...current(state) };
      newState.cars = action.payload.cars;
      newState.status = 'succeeded';
      return newState;
    },
    [getCars.rejected]: (state, action) => {
      const newState = { ...current(state) };
      newState.status = 'failed';
      newState.error = action.error.message;
      return newState;
    },
  },
});

export default carSlice.reducer;
