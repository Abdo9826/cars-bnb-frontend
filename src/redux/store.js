import { configureStore } from '@reduxjs/toolkit';
/* eslint-disable-next-line import/no-extraneous-dependencies */
import logger from 'redux-logger';
import CarsReducer from './Cars/carSlice';
import authenticationReducer from './Auth/authenticationSlice';
import ReserveReducer from './Reservations/ReserveSlice';
import reservationListReducer from './reservationList/reservationListSlice';

export default configureStore({
  reducer: {
    user: authenticationReducer,
    cars: CarsReducer,
    addReservations: ReserveReducer,
    reservations: reservationListReducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(logger),
});
