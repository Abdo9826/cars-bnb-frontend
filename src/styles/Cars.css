.main-car-container {
  background-color: #f7f7f7;
  display: flex;
  flex-direction: column;
  align-items: center;

  /* justify-content: center; */
}

.main-car-container > h1,
.main-car-container > p {
  font-family: 'Roboto', sans-serif;
  font-weight: 300;
  text-align: center;
  margin: 1rem 0;
  padding: 0;
}

.main-car-container > h1 {
  font-weight: 700;
}

.car-container {
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  align-items: center;
  margin: 0 auto;
  width: 100%;
}

.car-card {
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 320px;
  color: grey;
  height: 400px;
  margin: 10px;
  padding: 2rem 10px;

  /* border: 1px solid red; */
}

.car-image-wrap {
  width: 80%;
  height: 140px;
  margin-bottom: 1rem;
}

.car-image-wrap > img {
  width: 100%;
}

.car-card > h2 {
  font-size: 1rem;
  font-weight: 700;
  margin: 10px 0;
  color: rgb(0, 0, 0);
}

.car-card > p {
  text-align: center;
  color: grey;
}

.cars-icons-wrap {
  border: 1px solid var(--grey-shade);
  width: 30px;
  height: 30px;
  font-size: 1rem;
  border-radius: 50%;
  margin: 0 0.5rem;
}

.car-link {
  text-decoration: none;
}

.car-details-container {
  position: relative;
  flex-direction: column;
  column-gap: 1rem;
  display: flex;
  width: 100%;
  background-color: var(--cars-bg);
}

.car-details-image-wrap {
  display: flex;
  flex-direction: column;
  align-items: center;
  margin: 2rem 0 1rem;
}

.car-details-image-wrap > img {
  width: 100%;
}

.car-details {
  font-family: 'inter', sans-serif;
  padding: 0 1rem;
}

.car-details-back-btn {
  position: absolute;
  top: 0;
  left: 0;
  font-size: 1.5rem;
  border: none;
  background-color: transparent;
  cursor: pointer;
}

.car-details > h2 {
  font-size: 1.5rem;
  font-weight: 700;
  margin: 0;
}

.loading {
  display: flex;
  justify-content: center;
  align-items: center;
  color: var(--primary-color);
}

@media screen and (min-width: 768px) {
  .car-layout {
    display: grid;
    grid-template-columns: 250px 1fr;
  }

  .car-details-container {
    flex-direction: row;
  }

  .car-details-image-wrap {
    flex: 2 0 10%;
    flex-direction: row;
    height: 100%;
    margin: 0;
  }

  .car-details {
    padding: 4rem 0.5rem 2rem;
    flex: 1;
    margin: 4rem 0 2rem 0;
  }

  .car-details-back-btn {
    top: 20px;
    padding: 0 0.5rem 0.5rem;
    width: 70px;
    border-top-right-radius: 25px;
    border-bottom-right-radius: 25px;
    color: #fff;
    background-color: var(--primary-color);
    height: 60px;
  }
}
