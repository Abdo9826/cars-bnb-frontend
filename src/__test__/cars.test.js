import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import Cars from '../components/cars/Cars';
import store from '../redux/store';

it('To check test is working', () => {
  expect(true).toBe(true);
});

it('Test if header render correctly', () => {
  const TREE = renderer.create(
    <Provider store={store}>
      <BrowserRouter>
        <Cars />
      </BrowserRouter>
    </Provider>,
  );
  expect(TREE).toMatchSnapshot();
});
