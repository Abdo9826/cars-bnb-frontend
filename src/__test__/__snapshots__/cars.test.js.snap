// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test if header render correctly 1`] = `
<div
  className="main-car-container"
>
  <h1>
    Latest Models
  </h1>
  <p>
    Please select a car model
  </p>
  <div
    className="car-container"
  >
    <div
      className="loading"
    >
      <h4>
        No Cars to show, Please add new car
      </h4>
    </div>
  </div>
</div>
`;
